src/Testing/CodeHacking/CodeHacker.php

```diff
@@ -0,0 +1,528 @@
+			stream_wrapper_register( self::PROTOCOL, __CLASS__ );
+			self::$enabled = true;
+		}
+	}
+
+	/**
+	 * Disable the code hacker.
+	 */
+	public static function disable() {
+		if ( self::$enabled ) {
+			stream_wrapper_restore( self::PROTOCOL );
+			self::$enabled = false;
+		}
+	}
+
+	/**
+	 * Unregister all the non-persistent registered hacks.
+	 */
+	public static function clear_hacks() {
+		self::$hacks = self::$persistent_hacks;
+	}
+
+	/**
+	 * Check if the code hacker is enabled.
+	 *
+	 * @return bool True if the code hacker is enabled.
+	 */
+	public static function is_enabled() {
+		return self::$enabled;
+	}
+
+	/**
+	 * Check if persistent hacks have been registered.
+	 *
+	 * @return bool True if persistent hacks have been registered.
+	 */
+	public static function has_persistent_hacks() {
+		return count( self::$persistent_hacks ) > 0;
+	}
+
+	/**
+	 * Register a new hack.
+	 *
+	 * @param mixed $hack A function with signature "hack($code, $path)" or an object containing a method with that signature.
+	 * @param bool  $persistent If true, the hack will be registered as persistent (so that clear_hacks will not clear it).
+	 * @throws \Exception Invalid input.
+	 */
+	public static function add_hack( $hack, $persistent = false ) {
+		if ( ! is_callable( $hack ) && ! is_object( $hack ) ) {
+			throw new \Exception( "Hacks must be either functions, or objects having a 'process(\$text, \$path)' method." );
+		}
+
+		if ( ! self::is_valid_hack_callback( $hack ) && ! self::is_valid_hack_object( $hack ) ) {
+			throw new \Exception( "CodeHacker::addhack: hacks must be either a function with a 'hack(\$code,\$path)' signature, or an object containing a public method 'hack' with that signature. " );
+		}
+
+		if ( $persistent ) {
+			self::$persistent_hacks[] = $hack;
+		}
+		self::$hacks[] = $hack;
+	}
+
+	/**
+	 * Check if the supplied argument is a valid hack callback (has two mandatory arguments).
+	 *
+	 * @param mixed $callback Argument to check.
+	 *
+	 * @return bool true if the argument is a valid hack callback, false otherwise.
+	 * @throws ReflectionException Error when instantiating ReflectionFunction.
+	 */
+	private static function is_valid_hack_callback( $callback ) {
+		return is_callable( $callback ) && 2 === ( new ReflectionFunction( $callback ) )->getNumberOfRequiredParameters();
+	}
+
+	/**
+	 * Check if the supplied argument is a valid hack object (has a public "hack" method with two mandatory arguments).
+	 *
+	 * @param mixed $callback Argument to check.
+	 *
+	 * @return boolt rue if the argument is a valid hack object, false otherwise.
+	 */
+	private static function is_valid_hack_object( $callback ) {
+		if ( ! is_object( $callback ) ) {
+			return false;
+		}
+
+		$ro = new ReflectionObject( ( $callback ) );
+		try {
+			$rm = $ro->getMethod( 'hack' );
+			return $rm->isPublic() && ! $rm->isStatic() && 2 === $rm->getNumberOfRequiredParameters();
+		} catch ( ReflectionException $exception ) {
+			return false;
```

Comment:
See `add_hack` method: when this returns false an exception with a proper error message is thrown. Also this is a private method and isn't used anywhere else.

Reply:
[not a blocker] IMO it will be better to throw an error from as deep in the stack as possible and then bubble it up to the user (unless we need to mask the error which does not seem to be the case here).